[FEAT] Kakao 소셜 로그인 구현

Hminkyung opened the pull request:
## #️⃣ 연관된 이슈

> 관련된 이슈 번호를 적어주세요.\
> Close #20 

## ✨ 작업 내용 (Summary)

> 이번 PR에서 작업한 내용을 간략히 설명해주세요. (이미지 첨부 가능)
### 카카오 소셜 로그인을 진행하였습니다.

- hrr라는 이름의 앱을 kakao에 등록하고 Client key와 Secret key를 발급받았습니다.
- Redirect URI 와 웹 플랫폼으로 등록하였고(앱 플랫폼의 경우는 추후 프론트에서 작업이 끝난뒤에 연동할 때 추가하면 된다고 합니다)=> 백엔드에서 테스트하기 위해 웹 플랫폼으로 등록하였습니다.
- 카카오 개인정보에서 닉네임(필수), 프로필 사진(선택) 으로 클릭했을 때 로그인이 가능하도록 하였습니다.
- Auth domain을 생성해서 추후에 있을 다른 소셜로그인을 위해서 AuthController, AuthRequsetDto, AuthResponseDto, AuthService, JwtService를 공통으로 사용할 수 있게 구현하였고 kakao 방식과 동일하게 추후 생길 네이버 로그인과 애플 로그인은 KakaoTokenResponse, KakaoUserResponse, KakaoAuthService등으로 추가될 예정입니다.
- 카카오 로그인 부분이 추가되도록 application.yml, application-dev.yml, application-docker.yml, application-local.yml 부분에 코드를 추가했습니다. 각 코드에 적힌 앱 스킴 주소(hrr://oauth/kakao)는 추후에 앱과 연동될 때 사용될 것이며 그 전까지는 웹으로 사용되기 위해 http://localhost:8080/api/v1/auth/kakao/callback으로 사용합니다.
- .env 파일이 수정되어서 수정된 부분을 노션 공유문서에 있는 환경변수 페이지에 새로 업로드 해뒀으니 반드시 참고 바랍니다.

---

## ✅ 변경 사항 체크리스트

> 다음 항목들을 확인하고 체크해주세요.

- [x] 코드에 영향이 있는 모든 부분에 대한 테스트를 작성하고 실행했나요?
- [x] 문서를 작성하거나 수정했나요? (필요한 경우)
- [ ] 중요한 변경 사항이 팀에 공유되었나요?

---

## 🧪 테스트 결과

> 코드 변경에 대해 테스트를 수행한 결과를 요약해주세요.

* **테스트 환경:** 로컬
* **테스트 방법:** Postman, 웹 주소, Swagger
* **결과 요약:** 모든 테스트 통과

---

## 📸 스크린샷

> 관련된 스크린샷 또는 GIF가 있다면 여기에 첨부해주세요.
**진행과정**

- 아래는 앱 등록 과정입니다.

<img width="1405" height="1014" alt="스크린샷 2025-11-01 212245" src="https://github.com/user-attachments/assets/dcdf3ca0-b305-468e-af01-1b6bb7d973e5" />
<img width="1111" height="521" alt="image" src="https://github.com/user-attachments/assets/e2db5993-3d41-4c66-bc33-0ae3e75c3f8a" />
<img width="938" height="642" alt="image" src="https://github.com/user-attachments/assets/e6526c41-ba11-40dc-b4a1-4bedb58bc879" />
<img width="1052" height="856" alt="image" src="https://github.com/user-attachments/assets/f18cb1ce-a261-4235-9a5e-85dcefe6d203" />
<img width="1111" height="338" alt="image" src="https://github.com/user-attachments/assets/19fedd70-2b7a-4b2a-a3b5-4f3c9a931a8e" />

- 클라이언트 시크릿 키 발급

<img width="1088" height="377" alt="image" src="https://github.com/user-attachments/assets/c437508f-87ce-40c3-9c85-276baa36479a" />

- 앱 키 발급 : 여기서 Rest API 키가 클라이언트ID로 사용됩니다

<img width="1090" height="634" alt="image" src="https://github.com/user-attachments/assets/91837961-25a8-4f5e-b0ca-edb5b94480ef" />

**테스트 과정**

- 테스트를 진행할 때는 아래와 같은 순서로 이루어집니다.

1. 카카오 로그인 진행: 이때 웹 주소 URL은 
https://kauth.kakao.com/oauth/authorize?response_type=code&client_id={REST_API_KEY}&redirect_uri={REDIRECT_URI}
=> REST_API_KEY는 .env 파일에서의 KAKAO_CLIENT_ID 이고 REDIRECT_URI는 .env 파일에서의 KAKAO_REDIRECT_URI 입니다.

2. 접속하면 카카오 로그인이 순차적으로 진행됩니다. 아이디와 비밀번호를 입력하고 난 뒤에는 아래와 같이 동의를 클릭합니다.
<img width="910" height="787" alt="image" src="https://github.com/user-attachments/assets/3717accb-36ee-4940-abf9-7d44ab1287d2" />
<img width="830" height="425" alt="image" src="https://github.com/user-attachments/assets/3a4b170d-fbfd-4b6c-80ac-7153aee1509f" />
<img width="761" height="426" alt="image" src="https://github.com/user-attachments/assets/f2d4a758-de0e-4bf5-af31-359e2e227f12" />
<img width="751" height="396" alt="image" src="https://github.com/user-attachments/assets/0038bd3c-009e-4639-825e-3047f00c53c1" />

3. 로그인까지 완료되면 whiteLabel Error page가 뜨는데 이거는 프론트 화면이 구현되지 않은 상태라서 자연스러운 장면이고 이때 URL 주소에서 코드값을 받아옵니다. 코드값은 URL 뒤에  code = 뒤에 있는 값을 복사해줍니다. 
<img width="1120" height="639" alt="image" src="https://github.com/user-attachments/assets/abb903e7-dfdb-4adc-b89c-4cdb6c741a2f" />

4. 이 인가 코드값은 백엔드에서 받는 구조로 구현했어요(일회용이에요) 이 값을 통해 swagger랑 postman으로 테스트를 진행합니다.
<img width="844" height="857" alt="image" src="https://github.com/user-attachments/assets/e42c5daf-3c85-4cb3-8180-6bbb52827deb" />
<img width="687" height="863" alt="image" src="https://github.com/user-attachments/assets/e634ce45-5d0c-4ba9-b55a-8672b73e3f22" />

짜라란, token까지 잘 받아오는걸 확인할 수 있씁니다.
이제 데이터베이스를 볼까용?
<img width="1125" height="177" alt="image" src="https://github.com/user-attachments/assets/1e60ccba-b5ba-448c-b092-08ee4fd84a78" />
데이터베이스에도 민경씨가 잘 들어와있네여


---

## 💬 리뷰 요구사항

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요.
인가코드값이 백엔드에서도 받을 수 있고 프론트에서도 받을 수 있게 되어있어서 고민하다가 백엔드에서 받을 수 있게 구현했습니다.
만약에 프론트에서 받게 구현했다면 ReactNative에서 카카오 로그인 창 열어서 -> 카카오가 redirect_uri로 프론트 주소로 인가코드 전달하고 -> 프론트가 받은 code를 API 요청으로 백엔드에 넘기는 방식인데 이렇게 말고 백엔드 리디렉션 방식으로 구현했어요.
즉, ReactNative에서 카카오 로그인을 실행-> 하면 카카오 로그인 성공되고 -> 카카오가 인가코드를 백엔드 redirect_uri로 전달하게 하면 백엔드(/api/v1/auth/kakao/callback)로 토큰 교환되도록 구현해두었습니다.
**!! 그리고 반드시 수정된 .env 파일 복사해주세요!!**
---

## 📎 참고 자료

> 관련 문서, 레퍼런스 링크 등이 있다면 여기에 첨부해주세요.
https://data-jj.tistory.com/53
https://developers.kakao.com/docs/latest/ko/kakaologin/prerequisite
https://developers.kakao.com/docs/latest/ko/rest-api/getting-started
https://developers.kakao.com/docs/latest/ko/kakaologin/common
https://ddonghyeo.tistory.com/16
https://notavoid.tistory.com/82
https://velog.io/@devlyny/Spring-Security-JWT-%EC%B9%B4%EC%B9%B4%EC%98%A4-%EC%86%8C%EC%85%9C-%EB%A1%9C%EA%B7%B8%EC%9D%B8-%EA%B5%AC%ED%98%84
https://bcp0109.tistory.com/379


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * 카카오 소셜 로그인 추가 및 카카오 계정으로 가입/로그인 가능.
  * POST 소셜 로그인, 카카오 콜백, 토큰 재발급 엔드포인트 제공.
  * 닉네임·프로필 이미지 자동 동기화 및 온보딩 상태 관리(LoginStatus).
  * 액세스/리프레시 JWT 발급·검증·재발급 지원 및 JWT 인증 필터 적용.
  * Swagger/OpenAPI에 Bearer 인증 연동.
* **Chores**
  * 카카오 관련 환경 설정(클라이언트 ID/시크릿/리다이렉트 등) 및 보안/CORS 설정 추가.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Hminkyung committed:
feat: Kakao 소셜 로그인에서 로직 구현

Hminkyung committed:
feat: swagger로 형식 변환

Hminkyung committed:
feat: 소셜로그인 공통 코드로 묶기 및 불필요한 주석 코드 삭제

yangjiae12 commented on src/main/java/com/hrr/backend/domain/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,41 @@
+                    required = true,
+                    schema = @Schema(type = "string", allowableValues = {"kakao", "naver", "apple"}, example = "kakao")
+            )
+            @PathVariable("socialType") String socialType,
```
현재 소셜 타입을 `String`으로 받아서 문자열 비교를 하고 있는데, `SocialType` `enum`을 만들어서 관리하면 더 좋을 것 같습니다!

yangjiae12 commented on src/main/java/com/hrr/backend/domain/auth/service/AuthService.java:
```diff
@@ -0,0 +1,40 @@
+    public AuthResponseDto.LoginResponse socialLogin(String socialType, AuthRequestDto.SocialLoginRequest request) {
+        if (!socialType.equalsIgnoreCase("kakao")) {
+            throw new IllegalArgumentException("현재는 kakao 로그인만 지원합니다.");
+        }
```
에러 메시지를 서비스 단에 직접 작성하기보다는 기존에 정의된 `ErrorCode`에 해당 예외를 추가하여 예외를 던지면 더 일관성 있을 것 같습니다!

yangjiae12 reviewed (commented):
수고하셨습니다! 승인은 파트장님이 해주시는 걸로 하죠...

yc3697 reviewed (changes requested):
응답에 추가할 거
- 	로그인 상태(신규가입 or 기존 유저)
- 	userId

추가할 메소드(Spring Security 활용)
- 	토큰 만료 여부 확인
- 	토큰 재발급(RT 만료됐을 때 AT 재발급 메소드 포함)
- 	토큰 유효성 검사
- 	토큰에서 userId 추출
- 	HTTP 요청 헤더에서 JWT 추출
- 	SecurityContextHolder 활용하여 사용자 정보 로딩

위 사항들의 추가 구현이 필요할 것 같습니다!

Hminkyung committed:
Merge branch 'develop' of https://github.com/Hrr-HabbitRoutineResult/hrr-server into feat/20-kakao-login-setting
chore: merge develop into feat/20-kakao-login-setting

Hminkyung committed:
refactor: socialType을 enum으로 만들어서 관리하기

Hminkyung committed:
refactor: ErrorCode에 예외를 추가하여 에러 던지기

Hminkyung committed:
feat: 소셜로그인 응답에 userId와 loginStatus 추가

yc3697 commented:
@coderabbitai full review

yc3697 commented:
@coderabbitai full review

Hminkyung committed:
refactor: UTF-8로 바이트를 고정하고 key 기반 signWith 오버로드 사용하게 변경

Hminkyung committed:
refactor: 카카오 API 호출 실패 시 예외 처리 추가

Hminkyung committed:
refactor: WebClient에 타임아웃 설정

Hminkyung committed:
refactor: 에러처리 추가 및 블로킹 호출 개선하기

Hminkyung committed:
refactor: yml 파일 통일 및 클라이언트 시크릿에 공백 제거

Hminkyung committed:
feat: JWT 생성, 검증, 추출 구현하기

Hminkyung committed:
feat: 요청에서 Authorization 헤더를 검사하고 토큰이 유효할 때 인증 정보 주입하기

Hminkyung committed:
feat: Spring Security 활성화해서 swagger랑 카카오 인증 제외하고 인증 없이 못 들어가게 막기

Hminkyung committed:
Merge branch 'develop' of https://github.com/Hrr-HabbitRoutineResult/hrr-server into feat/20-kakao-login-setting

Hminkyung committed:
feat: 토큰 재발급 API 구현 및 인증 처리 에러 고치기

Hminkyung committed:
feat: 에러 수정 완료 및 소셜 로그인 구현 응답 처리 완료

Hminkyung committed:
feat: 프론트와 연동 및 스웨거로도 테스트 가능하게 로직 변경

yc3697 commented on src/main/java/com/hrr/backend/domain/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,61 @@
+    @Operation(summary = "소셜 로그인", description = "프론트에서 카카오 인가 코드를 받아 로그인 요청합니다.")
+    @PostMapping(value = "/social-login/{socialType}", produces = "application/json")
+    public ApiResponse<AuthResponseDto.LoginResponse> socialLogin(
+            @Parameter(
+                    description = "소셜 로그인 타입 (KAKAO, NAVER, APPLE)",
+                    required = true,
+                    schema = @Schema(type = "string", allowableValues = {"KAKAO", "NAVER", "APPLE"}, example = "KAKAO")
+            )
```
이 부분 아직 string으로 요청 받고 있어서 구현한 enum으로 변경 필요할 것 같아요

yc3697 commented on src/main/java/com/hrr/backend/domain/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,61 @@
+        AuthRequestDto.SocialLoginRequest request = new AuthRequestDto.SocialLoginRequest(code);
+        AuthResponseDto.LoginResponse response = authService.socialLogin(SocialType.KAKAO, request);
+        return ApiResponse.onSuccess(SuccessCode.OK, response);
+    }
```
string으로 code를 받고 requestDto를 내부에서 직접 생성하는 건 dto를 만들어놓은 의미가 없어보입니다. 
전달받은 code를 바로 socialLogin의 파라미터로 넘기거나, requestDto로 받아오게 하는 게 좋을 것 같아요

yc3697 commented on src/main/java/com/hrr/backend/global/config/SecurityConfig.java:
```diff
@@ -1,35 +1,89 @@
+    }
+
+    @Bean
+    public CorsConfigurationSource corsConfigurationSource() {
```
혹시 이건 적용되었을까요??

yc3697 commented on src/main/java/com/hrr/backend/global/response/ErrorCode.java:
```diff
@@ -23,7 +23,17 @@ public enum ErrorCode implements BaseCode{
 	DM_MESSAGE_NOT_FOUND(HttpStatus.NOT_FOUND, "DM404", "존재하지 않는 대화입니다."),
 
-	;
+    // auth
```
에러 코드 중복 해결해주세요!

yc3697 reviewed (approved):
수정 부탁드립니다~

Hminkyung committed:
refactor: 코드 리뷰 반영
-에러 코드 중복해결
-불필요한 주석 제거
-CORS 설정을 SecurityFilterChain에 적용
-requiredargsconstrucotr 어노테이션 제거
-AuthController에서 enum 요청으로 변경
-전달받은 code를 requestDto로 받아오게 하기

Hminkyung committed:
Resolve merge conflict in User.java

yc3697 reviewed (approved):
확인했습니다! 수고하셨어요~
